### PR TITLE
Ensure error is shown correctly for empty headers field

### PR DESCRIPTION
**Pull request by ijjk**
This ensures we show an error when an empty headers field is passed for a header item correctly. 

## Bug

- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [x] Errors have helpful link attached, see `contributing.md`

x-ref: https://github.com/vercel/vercel/discussions/6623

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| buildDuration | 16.3s | 15.6s | -631ms |
| buildDurationCached | 3.4s | 3.4s | ⚠️ +1ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | ⚠️ +317 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.83 | 2.905 | ⚠️ +0.07 |
| / avg req/sec | 883.52 | 860.71 | ⚠️ -22.81 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.731 | 1.866 | ⚠️ +0.14 |
| /error-in-render avg req/sec | 1443.98 | 1339.74 | ⚠️ -104.24 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| 745.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 23.3 kB | 23.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 67.1 kB | 67.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| _app-HASH.js gzip | 979 B | 979 B | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| dynamic-HASH.js gzip | 2.67 kB | 2.67 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 918 B | 918 B | ✓ |
| image-HASH.js gzip | 4.14 kB | 4.14 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 318 B | 318 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 13 kB | 13 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| _buildManifest.js gzip | 491 B | 491 B | ✓ |
| Overall change | 491 B | 491 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| index.html gzip | 539 B | 539 B | ✓ |
| link.html gzip | 552 B | 552 B | ✓ |
| withRouter.html gzip | 534 B | 534 B | ✓ |
| Overall change | 1.63 kB | 1.63 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| buildDuration | 13.1s | 12.9s | -245ms |
| buildDurationCached | 5.4s | 5.3s | -155ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | ⚠️ +317 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.115 | 3.072 | -0.04 |
| / avg req/sec | 802.69 | 813.74 | +11.05 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.901 | 1.81 | -0.09 |
| /error-in-render avg req/sec | 1315.11 | 1381.57 | +66.46 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| 16.HASH.js gzip | 186 B | 186 B | ✓ |
| 677f882d2ed8..HASH.js gzip | 14.1 kB | 14.1 kB | ✓ |
| framework.HASH.js gzip | 41.9 kB | 41.9 kB | ✓ |
| main-HASH.js gzip | 10.6 kB | 10.6 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 68 kB | 68 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| _app-HASH.js gzip | 964 B | 964 B | ✓ |
| _error-HASH.js gzip | 3.8 kB | 3.8 kB | ✓ |
| amp-HASH.js gzip | 552 B | 552 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| dynamic-HASH.js gzip | 2.87 kB | 2.87 kB | ✓ |
| head-HASH.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks-HASH.js gzip | 924 B | 924 B | ✓ |
| index-HASH.js gzip | 231 B | 231 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 298 B | 298 B | ✓ |
| script-HASH.js gzip | 3 kB | 3 kB | ✓ |
| withRouter-HASH.js gzip | 295 B | 295 B | ✓ |
| 30809af5c834..565.css gzip | 125 B | 125 B | ✓ |
| Overall change | 18.1 kB | 18.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| _buildManifest.js gzip | 501 B | 501 B | ✓ |
| Overall change | 501 B | 501 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| index.html gzip | 586 B | 586 B | ✓ |
| link.html gzip | 597 B | 597 B | ✓ |
| withRouter.html gzip | 578 B | 578 B | ✓ |
| Overall change | 1.76 kB | 1.76 kB | ✓ |

</details>

</details>
Commit: 7721a3ee9ee22dd7a650aafa66faf129995e8841

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7721a3ee9ee22dd7a650aafa66faf129995e8841

**test/acceptance/ReactRefreshLogBox.dev.test.js** 
-  > logbox: can recover from a syntax error without losing state

<details>
<summary>Expand output</summary>

  ● logbox: can recover from a syntax error without losing state

    expect(received).toBe(expected) // Object.is equality

    Expected: "Count: 1"
    Received: "1"

      46 |       import { useCallback, useState } from 'react'
      47 |
    > 48 |       export default function Index() {
         |        ^
      49 |         const [count, setCount] = useState(0)
      50 |         const increment = useCallback(() => setCount(c => c + 1), [setCount])
      51 |         return (

      at Object.<anonymous> (acceptance/ReactRefreshLogBox.dev.test.js:48:8)

</details>

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7721a3ee9ee22dd7a650aafa66faf129995e8841

**test/integration/gssp-ssr-change-reloading/test/index.test.js** 
- GS(S)P Server-Side Change Reloading > should not reload page when client-side is changed too GSP
- GS(S)P Server-Side Change Reloading > should update page when getStaticProps is changed only for /index
- GS(S)P Server-Side Change Reloading > should not reload page when client-side is changed too GSSP

<details>
<summary>Expand output</summary>

  ● GS(S)P Server-Side Change Reloading › should not reload page when client-side is changed too GSP

    Failed to replace content.

    Pattern: change me

    Content: import { useRouter } from 'next/router'

    export default function Gsp(props) {
      if (useRouter().isFallback) {
        return 'Loading...'
      }

      return (
        <>
          <p id="change">changed</p>
          <p id="props">{JSON.stringify(props)}</p>
        </>
      )
    }

    export const getStaticProps = async ({ params }) => {
      const count = 1

      if (params.post === 'second') {
        await new Promise((resolve) => setTimeout(resolve, 2000))
      }

      return {
        props: {
          count,
          params,
          random: Math.random(),
        },
      }
    }

    export const getStaticPaths = () => {
      /* eslint-disable-next-line no-unused-vars */
      const paths = 1

      return {
        paths: [{ params: { post: 'first' } }, { params: { post: 'second' } }],
        fallback: true,
      }
    }

      461 |     }
      462 |
    > 463 |     const newContent = currentContent.replace(pattern, newValue)
          |                       ^
      464 |     this.write(newContent)
      465 |   }
      466 |

      at File1.replace (lib/next-test-utils.js:463:23)
      at Object.<anonymous> (integration/gssp-ssr-change-reloading/test/index.test.js:39:14)

  ● GS(S)P Server-Side Change Reloading › should update page when getStaticProps is changed only for /index

    SyntaxError: Unexpected end of JSON input
        at JSON.parse (<anonymous>)

       98 |     await check(
       99 |       async () =>
    > 100 |         JSON.parse(await browser.elementByCss('#props').text()).count + '',
          |                            ^
      101 |       '1'
      102 |     )
      103 |   })

      at Object.<anonymous> (integration/gssp-ssr-change-reloading/test/index.test.js:100:28)

  ● GS(S)P Server-Side Change Reloading › should not reload page when client-side is changed too GSSP

    Failed to replace content.

    Pattern: change me

    Content: export default function Gssp(props) {
      return (
        <>
          <p id="change">changed</p>
          <p id="props">{JSON.stringify(props)}</p>
        </>
      )
    }

    export const getServerSideProps = ({ params }) => {
      const count = 1

      return {
        props: {
          count,
          params,
          random: Math.random(),
        },
      }
    }

      461 |     }
      462 |
    > 463 |     const newContent = currentContent.replace(pattern, newValue)
          |                       ^
      464 |     this.write(newContent)
      465 |   }
      466 |

      at File1.replace (lib/next-test-utils.js:463:23)
      at Object.<anonymous> (integration/gssp-ssr-change-reloading/test/index.test.js:124:14)

</details>

File: packages/next/lib/load-custom-routes.ts [hunk 1]
```diff
@@ -108,6 +108,8 @@ function checkHeader(route: Header): string[] {
 
   if (!Array.isArray(route.headers)) {
     invalidParts.push('`headers` field must be an array')
+  } else if (route.headers.length === 0) {
+    invalidParts.push('`headers` field can not be empty')
```
Comment (styfle) [inline packages/next/lib/load-custom-routes.ts:112]:
```suggestion
    invalidParts.push('`headers` field cannot be empty')
```

https://www.merriam-webster.com/words-at-play/cannot-vs-can-not-is-there-a-difference

File: test/integration/invalid-custom-routes/test/index.test.js [hunk 2]
```diff
@@ -25,6 +25,23 @@ const writeConfig = async (routes, type = 'redirects') => {
...
+          source: `/:path*`,
+          headers: [],
+        },
+      ],
+      'headers'
+    )
+    const stderr = await getStderr()
+
+    expect(stderr).toContain(
+      '`headers` field can not be empty for route {"source":"/:path*"'
```
Comment (styfle) [inline test/integration/invalid-custom-routes/test/index.test.js:41]:
```suggestion
      '`headers` field cannot be empty for route {"source":"/:path*"'
```

**Review by styfle — APPROVED** [review]
Thanks!

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| buildDuration | 14s | 13.7s | -245ms |
| buildDurationCached | 3.3s | 3.2s | -65ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | ⚠️ +313 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.685 | 2.455 | -0.23 |
| / avg req/sec | 930.96 | 1018.46 | +87.5 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.471 | 1.35 | -0.12 |
| /error-in-render avg req/sec | 1699.38 | 1851.7 | +152.32 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| 745.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 23.3 kB | 23.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 67.2 kB | 67.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| _app-HASH.js gzip | 979 B | 979 B | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| dynamic-HASH.js gzip | 2.67 kB | 2.67 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 918 B | 918 B | ✓ |
| image-HASH.js gzip | 4.14 kB | 4.14 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 318 B | 318 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 13 kB | 13 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| _buildManifest.js gzip | 491 B | 491 B | ✓ |
| Overall change | 491 B | 491 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| index.html gzip | 540 B | 540 B | ✓ |
| link.html gzip | 552 B | 552 B | ✓ |
| withRouter.html gzip | 534 B | 534 B | ✓ |
| Overall change | 1.63 kB | 1.63 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| buildDuration | 11.3s | 11.3s | ⚠️ +28ms |
| buildDurationCached | 4.8s | 4.7s | -109ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | ⚠️ +313 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.475 | 2.445 | -0.03 |
| / avg req/sec | 1009.96 | 1022.47 | +12.51 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.377 | 1.379 | 0 |
| /error-in-render avg req/sec | 1815.12 | 1813.22 | ⚠️ -1.9 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| 16.HASH.js gzip | 186 B | 186 B | ✓ |
| 677f882d2ed8..HASH.js gzip | 14.1 kB | 14.1 kB | ✓ |
| framework.HASH.js gzip | 41.9 kB | 41.9 kB | ✓ |
| main-HASH.js gzip | 10.7 kB | 10.7 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 68.1 kB | 68.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| _app-HASH.js gzip | 964 B | 964 B | ✓ |
| _error-HASH.js gzip | 3.8 kB | 3.8 kB | ✓ |
| amp-HASH.js gzip | 552 B | 552 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| dynamic-HASH.js gzip | 2.87 kB | 2.87 kB | ✓ |
| head-HASH.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks-HASH.js gzip | 924 B | 924 B | ✓ |
| index-HASH.js gzip | 231 B | 231 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 298 B | 298 B | ✓ |
| script-HASH.js gzip | 3.03 kB | 3.03 kB | ✓ |
| withRouter-HASH.js gzip | 295 B | 295 B | ✓ |
| 30809af5c834..565.css gzip | 125 B | 125 B | ✓ |
| Overall change | 18.1 kB | 18.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| _buildManifest.js gzip | 500 B | 500 B | ✓ |
| Overall change | 500 B | 500 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js fix/empty-headers | Change |
| - | - | - | - |
| index.html gzip | 585 B | 585 B | ✓ |
| link.html gzip | 597 B | 597 B | ✓ |
| withRouter.html gzip | 579 B | 579 B | ✓ |
| Overall change | 1.76 kB | 1.76 kB | ✓ |

</details>

</details>
Commit: 3ff7e8ba6929c228ff177888410e70f935989fbd

**Pull request merged**